chronicles.nim

```diff
@@ -103,10 +106,14 @@ when runtimeFilteringEnabled:
   import chronicles/topics_registry
   export setTopicState, setLogLevel, TopicState
 
-  proc topicStateIMPL(topicName: static[string]): ptr Topic =
-    var topic {.global.}: Topic = Topic(state: Normal, logLevel: NONE)
-    var dummy {.global.} = registerTopic(topicName, addr(topic))
-    return addr(topic)
+  proc topicStateIMPL(topicName: static[string]): Topic =
+    var topic: Topic
+    try:
+      topic = getTopicState(topicName)
+    except KeyError:
+      topic = Topic(state: Normal, logLevel: NONE)
+      registerTopic(topicName, topic)
+    return topic
```

Comment:
Sorry, I wanted to spend some time with the code myself to help you finish this, but I'll have to provide some comments only.

Please try to get back to the previous version of `topicStateIMPL`. This would allow you to greatly reduce the need for locking in the rest of the code. The simple idea of `topicStateIMPL` is that each topic receives a global variable with a statically known address and most of code that needs to access and manipulate the state of a particular topic can work with a simple pointer to this address. Please note that the state variable does not contain garbage collected types and manipulating the state is just flipping an enum value, which is perfectly thread-safe.

`setTopicState("literal_name", ...)` can be translated to a write to the known location - no locking required.

The only place where we need locking is the registration of new topics (which happens at the start of the program before any threads are spawned). The more dynamic API (`getTopicState`, `setTopicState(non-static string)`) was added only to simplify the way the `chronicle_tail` tool works (because the tail tool doesn't known the names of topics in advance). Only this "dynamic" API needs locking.

Reply:
> Please try to get back to the previous version of topicStateIMPL.

I don't think the old version is correct. All the corresponding entries in the global topic registry were overwritten with `Topic(state: Normal, logLevel: NONE)` instances, deleting previous non-default values for those topics.

> Please note that the state variable does not contain garbage collected types and manipulating the state is just flipping an enum value, which is perfectly thread-safe.

You can't just assume that with modern architectures. You either do explicit locking or you use atomic operations: https://github.com/nim-lang/Nim/blob/master/lib/system/atomics.nim

> setTopicState("literal_name", ...) can be translated to a write to the known location - no locking required.

Until it's executed at the same time from different threads. It's better to be defensive here and assume the worst about the users of your API.

Reply:
This is a logging library. Extremely precise cache-coherent behavior of `setTopicState` is not really that important. Eventual consistency is good enough.

Reply:
> Eventual consistency is good enough.

We were not getting it at all.

`topicsMatch()` received a list of `Topic` objects created by `topicStateIMPL()` - all of them having a `Normal` state and a log level of `NONE`. Then `topicsMatch()` checked their states and log levels like there could be any other values in there.

Somewhere along the line, the logic was scrambled.

Reply:
Found the commit that messed up the logic: https://github.com/status-im/nim-chronicles/commit/f929f3005c24ea759c4b818546fc4942b2e1c2c9#diff-7810f6d32e19cbc385dbe25d827e77f9L133

Reply:
With the old code, `topicsMatch` received pointers to global objects that were "created" at the begging of the program. If you were always seeing `NONE` and `Normal` values for their fields, this is another bug that needs to be investigated.